### PR TITLE
[hue] Check HTTPS connection (download of PEM certificate)

### LINKED ISSUE #13586: [hue] Binding broken until you update the new protocol setting to "http" (bridge V1)
I just installed snapshot 3143 containing the PR #11842 merged yesterday.
The binding does no more initialize properly:
```
09:32:18.362 [ERROR] [connection.HueTlsTrustManagerProvider] - An unexpected exception occurred - returning a TrustAllTrustManager: java.net.ConnectException: Connexion refusée (Connection refused)
org.openhab.core.io.net.http.PEMTrustManager$CertificateInstantiationException: java.net.ConnectException: Connexion refusée (Connection refused)
        at org.openhab.core.io.net.http.PEMTrustManager.getPEMCertificateFromServer(PEMTrustManager.java:220) ~[bundleFile:?]
        at org.openhab.core.io.net.http.PEMTrustManager.getInstanceFromServer(PEMTrustManager.java:131) ~[bundleFile:?]
        at org.openhab.core.io.net.http.PEMTrustManager.getInstanceFromServer(PEMTrustManager.java:116) ~[bundleFile:?]
        at org.openhab.binding.hue.internal.connection.HueTlsTrustManagerProvider.getTrustManager(HueTlsTrustManagerProvider.java:63) [bundleFile:?]
        at org.openhab.core.io.net.http.internal.ExtensibleTrustManagerImpl.addTlsTrustManagerProvider(ExtensibleTrustManagerImpl.java:220) [bundleFile:?]
        at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
        at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
        at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
        at org.apache.felix.scr.impl.inject.methods.BaseMethod.invokeMethod(BaseMethod.java:244) [bundleFile:?]
        at org.apache.felix.scr.impl.inject.methods.BaseMethod.access$500(BaseMethod.java:41) [bundleFile:?]
        at org.apache.felix.scr.impl.inject.methods.BaseMethod$Resolved.invoke(BaseMethod.java:685) [bundleFile:?]
        at org.apache.felix.scr.impl.inject.methods.BaseMethod.invoke(BaseMethod.java:529) [bundleFile:?]
        at org.apache.felix.scr.impl.inject.methods.BindMethod.invoke(BindMethod.java:42) [bundleFile:?]
        at org.apache.felix.scr.impl.manager.DependencyManager.doInvokeBindMethod(DependencyManager.java:2083) [bundleFile:?]
        at org.apache.felix.scr.impl.manager.DependencyManager.invokeBindMethod(DependencyManager.java:2058) [bundleFile:?]
        at org.apache.felix.scr.impl.manager.SingleComponentManager.invokeBindMethod(SingleComponentManager.java:443) [bundleFile:?]
        at org.apache.felix.scr.impl.manager.DependencyManager$MultipleDynamicCustomizer.addedService(DependencyManager.java:333) [bundleFile:?]
        at org.apache.felix.scr.impl.manager.DependencyManager$MultipleDynamicCustomizer.addedService(DependencyManager.java:301) [bundleFile:?]
        at org.apache.felix.scr.impl.manager.ServiceTracker$Tracked.customizerAdded(ServiceTracker.java:1200) [bundleFile:?]
        at org.apache.felix.scr.impl.manager.ServiceTracker$Tracked.customizerAdded(ServiceTracker.java:1121) [bundleFile:?]
        at org.apache.felix.scr.impl.manager.ServiceTracker$AbstractTracked.trackAdding(ServiceTracker.java:928) [bundleFile:?]
        at org.apache.felix.scr.impl.manager.ServiceTracker$AbstractTracked.track(ServiceTracker.java:864) [bundleFile:?]
        at org.apache.felix.scr.impl.manager.ServiceTracker$Tracked.serviceChanged(ServiceTracker.java:1152) [bundleFile:?]
        at org.apache.felix.scr.impl.BundleComponentActivator$ListenerInfo.serviceChanged(BundleComponentActivator.java:114) [bundleFile:?]
        at org.eclipse.osgi.internal.serviceregistry.FilteredServiceListener.serviceChanged(FilteredServiceListener.java:120) [org.eclipse.osgi-3.17.200.jar:?]
        at org.eclipse.osgi.internal.framework.BundleContextImpl.dispatchEvent(BundleContextImpl.java:956) [org.eclipse.osgi-3.17.200.jar:?]
        at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:234) [org.eclipse.osgi-3.17.200.jar:?]
        at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:151) [org.eclipse.osgi-3.17.200.jar:?]
        at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEventPrivileged(ServiceRegistry.java:936) [org.eclipse.osgi-3.17.200.jar:?]
        at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEvent(ServiceRegistry.java:873) [org.eclipse.osgi-3.17.200.jar:?]
        at org.eclipse.osgi.internal.serviceregistry.ServiceRegistrationImpl.register(ServiceRegistrationImpl.java:141) [org.eclipse.osgi-3.17.200.jar:?]
        at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.registerService(ServiceRegistry.java:261) [org.eclipse.osgi-3.17.200.jar:?]
        at org.eclipse.osgi.internal.framework.BundleContextImpl.registerService(BundleContextImpl.java:495) [org.eclipse.osgi-3.17.200.jar:?]
        at org.eclipse.osgi.internal.framework.BundleContextImpl.registerService(BundleContextImpl.java:514) [org.eclipse.osgi-3.17.200.jar:?]
        at org.openhab.binding.hue.internal.handler.HueBridgeHandler.initialize(HueBridgeHandler.java:713) [bundleFile:?]
        at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
        at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
        at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
        at org.openhab.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:154) [bundleFile:?]
        at org.openhab.core.internal.common.Invocation.call(Invocation.java:52) [bundleFile:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:834) [?:?]
Caused by: java.net.ConnectException: Connexion refusée (Connection refused)
        at java.net.PlainSocketImpl.socketConnect(Native Method) ~[?:?]
        at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:399) ~[?:?]
        at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:242) ~[?:?]
        at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:224) ~[?:?]
        at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:403) ~[?:?]
        at java.net.Socket.connect(Socket.java:609) ~[?:?]
        at sun.security.ssl.SSLSocketImpl.connect(SSLSocketImpl.java:289) ~[?:?]
        at sun.security.ssl.BaseSSLSocketImpl.connect(BaseSSLSocketImpl.java:173) ~[?:?]
        at sun.net.NetworkClient.doConnect(NetworkClient.java:182) ~[?:?]
        at sun.net.www.http.HttpClient.openServer(HttpClient.java:474) ~[?:?]
        at sun.net.www.http.HttpClient.openServer(HttpClient.java:569) ~[?:?]
        at sun.net.www.protocol.https.HttpsClient.<init>(HttpsClient.java:265) ~[?:?]
        at sun.net.www.protocol.https.HttpsClient.New(HttpsClient.java:372) ~[?:?]
        at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.getNewHttpClient(AbstractDelegateHttpsURLConnection.java:191) ~[?:?]
        at sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1187) ~[?:?]
        at sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1081) ~[?:?]
        at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:177) ~[?:?]
        at sun.net.www.protocol.https.HttpsURLConnectionImpl.connect(HttpsURLConnectionImpl.java:168) ~[?:?]
        at org.openhab.core.io.net.http.PEMTrustManager.getPEMCertificateFromServer(PEMTrustManager.java:209) ~[bundleFile:?]
        ... 45 more
```
The bridge thing remains in status UNKNOWN.

### PR DESCRIPTION
**Pull request by lolodomo**
Fix #13586

Signed-off-by: Laurent Garnier <lg.hc@free.fr>

### CODE REVIEW

**Comment by cweitkamp** [comment]
First of all thank you for taking care.

I see two issues with your approach. First one is that the certificate is requested/ loaded twice - which is the minor one. Second the constructor of the `HueBridge` already starts an authentication request which obviously will fail is HTTPS is activated because of the late tls provider registration.

**Comment by lolodomo** [comment]
> First one is that the certificate is requested/ loaded twice - which is the minor one.

Yes, that is true but that is only at the startup of OH (binding).

> Second the constructor of the `HueBridge` already starts an authentication request which obviously will fail is HTTPS is activated because of the late tls provider registration.

I don't see any request started. Here is the code of the called HueBridge contructor:
https://github.com/openhab/openhab-addons/blob/main/bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/connection/HueBridge.java#L128
Only class members are initialized.

**Comment by lolodomo** [comment]
Maybe you thought the second constructor was called, the one which also calls `authenticate`. But that is not the case.

**Comment by lolodomo** [comment]
> > First one is that the certificate is requested/ loaded twice - which is the minor one.
> 
> Yes, that is true but that is only at the startup of OH (binding).

That being said, I can avoid it.

**Comment by lolodomo** [comment]
> > > First one is that the certificate is requested/ loaded twice - which is the minor one.
> > 
> > 
> > Yes, that is true but that is only at the startup of OH (binding).
> 
> That being said, I can avoid it.

It is now fixed.

**Comment by lolodomo** [comment]
I cached the PEM trust manager to not download the certificate twice.
In case it should not be cached on a long period, I could replace the cache by an expiry cache. Just let me know.

File: bundles/org.openhab.binding.hue/src/main/resources/OH-INF/i18n/hue.properties [hunk 1]
```diff
@@ -148,6 +148,7 @@ config-status.error.missing-ip-address-configuration = No IP address for the Hue
 # thing status descriptions
 
 offline.communication-error = An unexpected exception occurred during execution.
+offline.conf-error-https-connection = HTTPS secure connection failed. Please check your configuration settings (network address, protocol, port, type of certificate) and change configuration to HTTP when using a V1 bridge.
```
Comment (jlaur) [inline bundles/org.openhab.binding.hue/src/main/resources/OH-INF/i18n/hue.properties:151]: Single comment, would prefer if @cweitkamp could join as reviewer:
```suggestion
offline.conf-error-https-connection = HTTPS secure connection failed. Please check your configuration settings (network address, protocol, port, type of certificate) and change protocol to http when using a V1 bridge.
```

Comment (lolodomo) [reply bundles/org.openhab.binding.hue/src/main/resources/OH-INF/i18n/hue.properties]: Done

**Comment by lolodomo** [comment]
@cweitkamp @jlaur : do not hesitate to test it in case you think it cannot work with a V2 bridge. I am confident it will work perfectly ;)

**Comment by jlaur** [comment]
> do not hesitate to test it in case you think it cannot work with a V2 bridge. I am confident it will work perfectly ;)

Already running it. :-)
- HTTPS communication still works (with default parameters).
- Setting protocol to `http` works.
- Setting `ipAddress` to some other host responding on port 443 but not exposing Hue API no longer causes infinite errors.
- Setting port to `80` and protocol to `https` causes same behavior as you probably see with Hue bridge V1. However, besides setting the bridge offline description, an exception is also logged with log level **ERROR**. Not sure if desired now? See below.

Exception:
```
2022-10-29 16:49:01.847 [ERROR] [onnection.HueTlsTrustManagerProvider] - An unexpected exception occurred: javax.net.ssl.SSLException: Unsupported or unrecognized SSL message
org.openhab.core.io.net.http.PEMTrustManager$CertificateInstantiationException: javax.net.ssl.SSLException: Unsupported or unrecognized SSL message
	at org.openhab.core.io.net.http.PEMTrustManager.getPEMCertificateFromServer(PEMTrustManager.java:220) ~[?:?]
	at org.openhab.core.io.net.http.PEMTrustManager.getInstanceFromServer(PEMTrustManager.java:131) ~[?:?]
	at org.openhab.core.io.net.http.PEMTrustManager.getInstanceFromServer(PEMTrustManager.java:116) ~[?:?]
	at org.openhab.binding.hue.internal.connection.HueTlsTrustManagerProvider.getPEMTrustManager(HueTlsTrustManagerProvider.java:77) ~[?:?]
	at org.openhab.binding.hue.internal.handler.HueBridgeHandler.lambda$10(HueBridgeHandler.java:721) ~[?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: javax.net.ssl.SSLException: Unsupported or unrecognized SSL message
	at sun.security.ssl.SSLSocketInputRecord.handleUnknownRecord(SSLSocketInputRecord.java:451) ~[?:?]
	at sun.security.ssl.SSLSocketInputRecord.decode(SSLSocketInputRecord.java:175) ~[?:?]
	at sun.security.ssl.SSLTransport.decode(SSLTransport.java:111) ~[?:?]
	at sun.security.ssl.SSLSocketImpl.decode(SSLSocketImpl.java:1506) ~[?:?]
	at sun.security.ssl.SSLSocketImpl.readHandshakeRecord(SSLSocketImpl.java:1416) ~[?:?]
	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:456) ~[?:?]
	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:427) ~[?:?]
	at sun.net.www.protocol.https.HttpsClient.afterConnect(HttpsClient.java:572) ~[?:?]
	at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:201) ~[?:?]
	at sun.net.www.protocol.https.HttpsURLConnectionImpl.connect(HttpsURLConnectionImpl.java:168) ~[?:?]
	at org.openhab.core.io.net.http.PEMTrustManager.getPEMCertificateFromServer(PEMTrustManager.java:209) ~[?:?]
	... 10 more
```

Perhaps it could be reduced to **DEBUG**?

**Comment by lolodomo** [comment]
> Perhaps it could be reduced to **DEBUG**?

OK

**Comment by lolodomo** [comment]
> * Setting `ipAddress` to some other host responding on port 443 but not exposing Hue API no longer causes infinite errors.

So I even fixed a bug I was not aware :)

**Comment by lolodomo** [comment]
> * Setting port to `80` and protocol to `https` causes same behavior as you probably see with Hue bridge V1. However, besides setting the bridge offline description, an exception is also logged with log level **ERROR**.

Fixed => DEBUG level now.

**Comment by cweitkamp** [comment]
> I cached the PEM trust manager to not download the certificate twice.
In case it should not be cached on a long period, I could replace the cache by an expiry cache. Just let me know.

There is a small chance that it will not work. E.g. if the Hue Bridge doesn't have a fixed IP address. But for now I'm happy with it.

I already know that there's a but inside the `PEMTrustManager` logic. The whole setup will fail once the certificate expires. This must be considered in the future.

**Comment by lolodomo** [comment]
So can we merge the PR ?

**Review by cweitkamp — APPROVED** [review]
LGTM. Thanks.

File: bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/HueBridgeHandler.java [hunk 2]
```diff
@@ -706,20 +706,35 @@ public void initialize() {
...
                 updateStatus(ThingStatus.UNKNOWN);
+
+                if (HueBridgeConfig.HTTPS.equals(hueBridgeConfig.protocol)) {
+                    scheduler.submit(() -> {
+                        // register trustmanager service
+                        HueTlsTrustManagerProvider tlsTrustManagerProvider = new HueTlsTrustManagerProvider(
+                                ip + ":" + hueBridgeConfig.getPort(), hueBridgeConfig.useSelfSignedCertificate);
+
+                        // Check before registering that the PEM certificate can be downloaded
+                        if (tlsTrustManagerProvider.getPEMTrustManager() == null) {
```
Comment (jlaur) [inline bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/HueBridgeHandler.java:721]: Just a heads up, will analyze and create issue later: After upgrading to RC1 I kept an eye on my things while booting, and I ended up in this situation. So I will need to analyze exactly how it happened - but it did, and as a result, no further attempts were made to get online and I had to manually restart the binding - after which everything worked fine.

**Pull request merged**